prov/efa/src/efa_av.c

```diff
@@ -185,60 +368,87 @@ struct efa_conn *efa_conn_alloc(struct efa_av *av, struct efa_ep_addr *raw_addr,
 		return NULL;
 	}
 
-	conn = ofi_ibuf_alloc(av->conn_pool);
-	if (!conn) {
-		EFA_WARN(FI_LOG_AV, "efa conn pool exhausted!\n");
-		errno = FI_ENOMEM;
+	err = ofi_av_insert_addr(&av->util_av, raw_addr, &util_av_fi_addr);
+	if (err) {
+		EFA_WARN(FI_LOG_AV, "ofi_av_insert_addr failed! Error message: %s\n",
+			 fi_strerror(err));
 		return NULL;
 	}
 
-	ibv_ah_attr.port_num = 1;
-	ibv_ah_attr.is_global = 1;
-	memcpy(ibv_ah_attr.grh.dgid.raw, raw_addr->raw, sizeof(raw_addr->raw));
-	conn->ah.ibv_ah = ibv_create_ah(ibv_pd, &ibv_ah_attr);
-	if (!conn->ah.ibv_ah) {
-		errno = FI_EINVAL;
-		goto err_free_conn;
-	}
+	util_av_entry = ofi_bufpool_get_ibuf(av->util_av.av_entry_pool,
+					     util_av_fi_addr);
+	efa_av_entry = (struct efa_av_entry *)util_av_entry->data;
+	conn = &efa_av_entry->conn;
 
-	memcpy((void *)&conn->ep_addr, raw_addr, sizeof(*raw_addr));
-	err = -efadv_query_ah(conn->ah.ibv_ah, &efa_ah_attr, sizeof(efa_ah_attr));
-	if (err) {
-		errno = err;
-		goto err_destroy_ah;
+	memcpy(&conn->ep_addr, raw_addr, sizeof(*raw_addr));
+	assert(av->type == FI_AV_MAP || av->type == FI_AV_TABLE);
+	conn->fi_addr = (av->type == FI_AV_MAP) ? (uintptr_t)(void *)conn : util_av_fi_addr;
+	conn->util_av_fi_addr = util_av_fi_addr;
+
+	conn->ah = efa_ah_alloc(av, raw_addr->raw);
+	if (!conn->ah)
+		goto err_release;
+
+	if (av->ep_type == FI_EP_RDM) {
+		err = efa_conn_rdm_init(av, conn);
+		if (err) {
+			errno = -err;
+			goto err_release;
+		}
 	}
 
-	conn->ah.ahn = efa_ah_attr.ahn;
-	key.ahn = conn->ah.ahn;
+	key.ahn = conn->ah->ahn;
 	key.qpn = raw_addr->qpn;
-	/* This is correct since the same address should be mapped to the same ah. */
-	HASH_FIND(hh, av->reverse_av, &key, sizeof(key), reverse_av);
-	if (!reverse_av) {
-		reverse_av = malloc(sizeof(*reverse_av));
-		if (!reverse_av) {
-			errno = FI_ENOMEM;
-			goto err_destroy_ah;
-		}
+	reverse_av_entry = NULL;
+	HASH_FIND(hh, av->reverse_av, &key, sizeof(key), reverse_av_entry);
```

Comment:
Can you tell me where this hash (av->reverse_av) is initialized to NULL per https://troydhanson.github.io/uthash/userguide.html#_declare_the_hash ?

Reply:
Also, a potential bug (and a number of Coverity/scan-build complaints) come from using a complex (struct) key field without memset-ing it to 0.  Here, for example, line 399 should contain a line memset-ing key to 0 (before initializing its subfields). https://troydhanson.github.io/uthash/userguide.html#_structure_keys

Reply:
@jagunnels

I just checked, `av->reverse_av` is initialized to NULL in `efa_av_open`. In that function, av was allocated as

` av = calloc(1, sizeof(*av));`

`calloc` will set the context to 0.

Reply:
Thanks.  The same comment applies to all hashes.  And the Coverity scan recently posted in https://github.com/ofiwg/libfabric/issues/6817.  This *could* be at the base of all of the division-by-zero flagged errors related to HASH_ADD, etc.  Zeroing out the contents of the structs used as keys appears to be the root cause of the Coverity scans that flag "The left operand of '<<' is a garbage value."  See https://troydhanson.github.io/uthash/userguide.html#_structure_keys

Reply:
I added a separate commit to zero out the content of key before calling `HASH_FIND/ADD`